src/backend/distributed/transaction/distributed_transaction_management.c

```diff
@@ -0,0 +1,431 @@
+				currentBackend->transactionId.timestamp;
+
+			SpinLockRelease(&currentBackend->mutex);
+
+			activeTransactionList = lappend(activeTransactionList, activeBackend);
+		}
+	}
+
+	return activeTransactionList;
+}
+
+
+/*
+ * InitializeDistributedTransactionManagement requests the necessary shared memory
+ * from Postgres and sets up the shared memory startup hook.
+ */
+void
+InitializeDistributedTransactionManagement(void)
+{
+	/* allocate shared memory */
+	RequestAddinShmemSpace(DistributedTransactionManagementShmemSize());
+
+	prev_shmem_startup_hook = shmem_startup_hook;
+	shmem_startup_hook = DistributedTransactionManagementShmemInit;
+}
+
+
+/*
+ * DistributedTransactionManagementShmemInit is the callback that is to be called on
+ * shared memory startup hook. The function sets up the necessary shared memory segment
+ * for the distributed transaction manager.
+ */
+static void
+DistributedTransactionManagementShmemInit(void)
+{
+	bool alreadyInitialized = false;
+	char *trancheName = "Distributed Transaction Management";
+
+	/* we may update the shmem, acquire lock exclusively */
+	LWLockAcquire(AddinShmemInitLock, LW_EXCLUSIVE);
+
+	distributedTransactionShmemData =
+		(DistributedTransactionShmemData *) ShmemInitStruct(
+			"Distributed Transaction Management",
+			DistributedTransactionManagementShmemSize(),
+			&alreadyInitialized);
+
+	if (!alreadyInitialized)
+	{
+		int backendIndex = 0;
+
+#if (PG_VERSION_NUM >= 100000)
+		NamedLWLockTranche *namedLockTranche =
+			&distributedTransactionShmemData->namedLockTranche;
+#else
+		LWLockTranche *lockTranche = &distributedTransactionShmemData->lockTranche;
+#endif
+
+		/* start by zeroing out all the memory */
+		memset(distributedTransactionShmemData, 0,
+			   DistributedTransactionManagementShmemSize());
+
+#if (PG_VERSION_NUM >= 100000)
+		namedLockTranche->trancheId = LWLockNewTrancheId();
+
+		LWLockRegisterTranche(namedLockTranche->trancheId, trancheName);
+		LWLockInitialize(&distributedTransactionShmemData->lock,
+						 namedLockTranche->trancheId);
+#else
+		distributedTransactionShmemData->trancheId = LWLockNewTrancheId();
+
+		/* we only need a single lock */
+		lockTranche->array_base = &distributedTransactionShmemData->lock;
+		lockTranche->array_stride = sizeof(LWLock);
+		lockTranche->name = trancheName;
+
+		LWLockRegisterTranche(distributedTransactionShmemData->trancheId, lockTranche);
+		LWLockInitialize(&distributedTransactionShmemData->lock,
+						 distributedTransactionShmemData->trancheId);
+#endif
+
+		/* start the distributed transaction ids from 1 */
+		pg_atomic_init_u64(&distributedTransactionShmemData->nextTransactionId, 1);
+
+		/*
+		 * We should init per backend's spinlock before any backend
+		 * starts its execution.
+		 */
+		for (backendIndex = 0; backendIndex < MaxBackends; ++backendIndex)
+		{
+			slock_t backendMutex =
+				distributedTransactionShmemData->sessions[backendIndex].mutex;
+
+			SpinLockInit(&backendMutex);
+		}
```

Comment:
This looks quite wrong - you're *copying* the spinlock into local memory, and then you're initializing that copy. IOW, this does't do anything lasting.